conda_index/index/__init__.py

```diff
@@ -743,17 +731,54 @@ def extract_wrapper(args):
                         log.debug("%s write index.html", subdir)
                         self._write_subdir_index_html(subdir, patched_repodata)
 
-                        t2.set_description("Updating channeldata")
-                        t2.update()
-                        log.debug("%s update channeldata", subdir)
-                        self._update_channeldata(channel_data, patched_repodata, subdir)
-
                         log.debug("%s finish", subdir)
 
-                # Step 7. Create and write channeldata.
-                self._write_channeldata_index_html(channel_data)
-                log.debug("write channeldata")
-                self._write_channeldata(channel_data)
+    def update_channeldata(self):
+        """
+        Update channeldata based on re-reading output `repodata.json` and existing
+        `channeldata.json`. Call after index() if channeldata is needed.
+        """
+        subdirs = self.detect_subdirs()
+
+        # Step 1. Lock local channel.
+        with utils.try_acquire_locks([utils.get_lock(self.channel_root)], timeout=900):
+            # keeep channeldata in memory, update with each subdir
+            channel_data = {}
+            channeldata_file = self.channeldata_path()
+            if os.path.isfile(channeldata_file):
+                with open(channeldata_file) as f:
+                    channel_data = json.load(f)
+
+            for subdir in subdirs:
+                log.info("Channeldata subdir: %s" % subdir)
+                log.debug("%s read repodata")
+                with open(
+                    os.path.join(self.output_root, subdir, REPODATA_JSON_FN)
+                ) as repodata:
+                    patched_repodata = json.load(repodata)
+
+                self._update_channeldata(channel_data, patched_repodata, subdir)
+
+                log.debug("%s channeldata finished", subdir)
+
+            # Step 7. Create and write channeldata.
+            self._write_channeldata_index_html(channel_data)
+            log.debug("write channeldata")
+            self._write_channeldata(channel_data)
+
+    def detect_subdirs(self):
+        if not self._subdirs:
+            detected_subdirs = {
+                subdir.name
+                for subdir in os.scandir(self.channel_root)
+                if subdir.name in utils.DEFAULT_SUBDIRS and subdir.is_dir()
+            }
+            log.debug("found subdirs %s" % detected_subdirs)
+            self.subdirs = sorted(detected_subdirs | {"noarch"})
+        else:
+            self.subdirs = sorted(set(self._subdirs))
+            log.warn("Indexing %s does not include 'noarch'", self.subdirs)
+        return self.subdirs
```

Comment:
```suggestion
    def _detect_subdirs(channel_root):
        detected_subdirs = {
            subdir.name
            for subdir in os.scandir(channel_root)
            if subdir.name in utils.DEFAULT_SUBDIRS and subdir.is_dir()
        }
        log.debug("found subdirs %s" % detected_subdirs)
        return sorted(detected_subdirs | {"noarch"})
```

Reply:
I do like moving functions that don't reference self out of the class.